accepted/2021/workloads/workload-target-imports.md

```diff
@@ -0,0 +1,118 @@
+```
+
+#### Pros:
+
+- Simple to implement
+- Low compat and performance impacts
+- Matches existing `BeforeTargetFrameworkInferenceTargets` property
+
+#### Cons:
+
+- Doesn’t match existing `Directory.Build.props` and `Directory.Build.targets` pattern, or the general principle of the SDK which is to have sensible convention-based defaults that can be overridden
+
+### New Automatically Imported .targets File
+
+Automatically find and import a `Directory.AfterTargetFrameworkInference.targets` file.
+
+#### Pros:
+
+- Matches existing convention for `Directory.Build.props` and `Directory.Build.targets`
+- Probably closest to what developers would expect
+
+#### Cons:
+
+- Possible performance impact
+- Adds another way for a build to “leak out” of a repo root (which could be a security issue)
+- More extension points may be needed in the future, and we probably don’t want to add a new auto-imported file for each one
+
+A possible mitigation for some of the cons could be to say that we don’t look for the new file to import independently. Rather, we could say that you need to have a `Directory.Build.props` file and it needs to be in the same folder as `Directory.Build.props`. So a `Directory.AfterTargetFrameworkInference.targets` file outside the repo root wouldn’t automatically be imported unless a `Directory.Build.props` file was going to be imported from outside the repo anyway.
+
+### Change import location of `Directory.Build.targets`
+
+Change`Directory.Build.targets` to be imported after TargetFramework parsing but before workloads and most of the .NET SDK and common targets. This would be a big breaking change but might not affect most people who use `Directory.Build.targets`. It would break people who use `Directory.Build.targets` to override properties or targets in the MSBuild common targets.
+
+#### Pros:
+
+- Logic to enable workloads in `Directory.Build.targets` would “just work”, which is likely what developers expect
+
+#### Cons
+
+- This would be a breaking change, possibly a very big one
+- Implementation would be somewhat complicated:
+  - Logic would be split betweenthe .NET SDK and MSBuild, as the .NET SDK would now be responsible for importing `Directory.Build.targets` for projects that use the .NET SDK, but other projects would still need MSBuild to import `Directory.Build.targets`
+  - The new behavior would need to be behind a change wave and ideally a separate opt-out flag
+- This goes against the original intent behind `Directory.Build.targets`, which was that it would be imported after all built-in targets.  However, the .NET SDK (or any MSBuild SDK) already break this.
+  - As part of this work we could possibly add another extension point that allows you to specify `.targets` files that are imported after all other projects or imports.  This would likely be an MSBuild engine feature.
```

Comment:
It would be good if you could state a preference from your team as you're the experts.

Reply:
Would it be at all possible to give users a bit more control over shimming each of these stages? E.g. in addition to Directory.Build.targets, maybe also Directory.Build.AfterSdks.targets and Directory.Build.BeforeSdk.targets. Sure, you'll still probably break things when you change things around, but then at least users that relied on the order would have a destination to move their definitions to.